checker/src/org/checkerframework/checker/lock/LockVisitor.java

```diff
@@ -353,9 +533,315 @@ protected boolean skipContractCheck(Tree tree, FlowExpressions.Receiver expr, Fl
+            }
+        }
+        
+        // Set any indices of 0 to -1
+        
+        for(int i = 0; i < guardSatisfiedIndex.length; i++) {
+            if (guardSatisfiedIndex[i] == 0) {
+                guardSatisfiedIndex[i] = -1;
+            }
+        }
+
+        // Combine all of the actual parameters into one list of AnnotationMirrors
+
+        ArrayList<AnnotationMirror> passedArgAnnotations = new ArrayList<AnnotationMirror>(guardSatisfiedIndex.length); // Not necessary to pass guardSatisfiedIndex.length, but it is known, so why not.
+        passedArgAnnotations.add(methodCallReceiver == null ? null : methodCallReceiver.getAnnotationInHierarchy(GUARDEDBYINACCESSIBLE));
+        for(ExpressionTree tree : node.getArguments()) {
+            passedArgAnnotations.add(atypeFactory.getAnnotatedType(tree).getAnnotationInHierarchy(GUARDEDBYINACCESSIBLE));
+        }
+
+        // Perform the validity check and issue an error if not valid.
+
+        for (int i = 0; i < guardSatisfiedIndex.length; i++) {
+            if (guardSatisfiedIndex[i] != -1) {
+                for (int j = i + 1; j < guardSatisfiedIndex.length; j++) {
+                    if (guardSatisfiedIndex[i] == guardSatisfiedIndex[j]) {
+                        // The @GuardedBy annotations must be identical on the corresponding actual parameters.
+                        AnnotationMirror anno1 = passedArgAnnotations.get(i);
+                        AnnotationMirror anno2 = passedArgAnnotations.get(j);
+                        if (anno1 != null && anno2 != null) {
+                            if (!atypeFactory.getQualifierHierarchy().isSubtype(anno1, anno2) ||
+                                !atypeFactory.getQualifierHierarchy().isSubtype(anno2, anno1)) {
+                                // TODO: allow these strings to be localized
+
+                                String formalParam1 = null;
+
+                                if (i == 0) {
+                                    formalParam1 = "The receiver type";
+                                } else {
+                                    formalParam1 = "Parameter #" + i; // i, not i-1, so the index is 1-based
+                                }
+
+                                String formalParam2 = "parameter #" + j; // j, not j-1, so the index is 1-based
+
+                                checker.report(Result.failure(
+                                        "guardsatisfied.parameters.must.match",
+                                        formalParam1, formalParam2, invokedMethod.toString(), guardSatisfiedIndex[i], anno1, anno2), node);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        return super.visitMethodInvocation(node, p);
+    }
+
+    // When visiting a synchronized block, issue an error if the expression
+    // has a type that implements the java.util.concurrent.locks.Lock inteface.
```

Comment:
This seems similar to the issue about 
>Look for special-case behavior for methods in checkers other than reflection and intents. For each of those, define and use a declaration annotation.

(Do we have a GitHub issue for this?)
Future enhancement might be to make a type declaration annotation for this rather than using Lock.class explicitly.